### PR TITLE
chore: add vscode workspace settings

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Share goodies with all the team.

**Description**

- As per VSCode `.vscode/settings.json` file is the only way to share single root workspace settings among team. 
- If any team member been using some good settings, would be nice to share with others

**Steps to test or reproduce**

- Run all tests 

**Additional Notes**

From the [VSCode](https://code.visualstudio.com/docs/getstarted/settings#_workspace-settings)

> VS Code stores workspace settings at the root of the project in a .vscode folder. This makes it easy to share settings with others in a version-controlled (for example, Git) project.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  },
+  "[typescript]": {
+    "editor.defaultFormatter": "biomejs.biome"
+  },
+  "[json]": {
+    "editor.defaultFormatter": "biomejs.biome" 
+  },
+  "[jsonc]": {
+    "editor.defaultFormatter": "biomejs.biome" 
+  }
```
Comment (nflaig) [inline .vscode/settings.json:21]: would be nice to include `"openInGitHub.defaultBranch": "unstable"` for [Open In GitHub](https://marketplace.visualstudio.com/items?itemName=sysoev.vscode-open-in-github) extension

**Review by matthewkeil — DISMISSED** [review]
Not sure how others will feel but im open to it.

**Comment by nazarhussain** [comment]
> Not sure how others will feel but im open to it.

Opened PR, to engage others in this discussion to get their opinion.

File: .vscode/extensions.json [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
   "recommendations": [
     "biomejs.biome",
-    "esbenp.prettier-vscode"
+    "esbenp.prettier-vscode",
+    "sysoev.vscode-open-in-github"
```
Comment (nflaig) [inline .vscode/extensions.json]: not sure if should go as far and add this here, might be better to only list essential extensions here

Comment (nazarhussain) [reply .vscode/extensions.json]: Without adding that extension, it's settings `openInGitHub.defaultBranch` will not be available. Also it's a recommended list, not compulsory for everyone.

Comment (matthewkeil) [reply .vscode/extensions.json]: not sure either.  there are a few that are super helpful like "npm intellisense" and "path intellisense" but its opening a can of worms...

Comment (nazarhussain) [reply .vscode/extensions.json]: @matthewkeil If you think any extension is useful for other team members and community who is contributing to this project, please share.

Comment (matthewkeil) [reply .vscode/extensions.json]: honestly i hate when repos bring up warning when i dont have an extension installed. its really annoying...i think only ones that are ABSOLUTELY necessary, like for the lint command, that should be included

Comment (nazarhussain) [reply .vscode/extensions.json]: For that I would recommend VSCode to have two workspace settings file. One `.vsode/settings.json` to shared among users and other `.vscode/settings.local.json` only for individual use. The only reason we can come up more settings in one file, is the different preferences of people.

Comment (nazarhussain) [reply .vscode/extensions.json]: @matthewkeil @nflaig So should I keep this plugin and related settings? or remove it from this PR?

Comment (nflaig) [reply .vscode/extensions.json]: > repos bring up warning when i dont have an extension installed.

that's my concern as well, not sure how often vs code will show the popup but better to keep this list to essential extensions

> So should I keep this plugin and related settings?

Keep the setting, remove the recommendation. Could add a comment in the setting file for which extension it is, people can manually install it if they want to

Comment (nazarhussain) [reply .vscode/extensions.json]: @nflaig Can't use comments for the `.json` files, and as per my knowledge VSCode not yet supporting `.jsonc` format.

Comment (matthewkeil) [reply .vscode/extensions.json]: Sounds good.  For opening in github and creating links to code i use gitlens as an example... i do not want to have to install that one

Comment (nflaig) [reply .vscode/extensions.json]: > Can't use comments for the .json files,

while the files are json and shouldn't allow for comments, you can still do it and vscode won't complain, it also does not show an error in the file itself but would do so for any other json file

**Comment by nazarhussain** [comment]
The right solution to this problem is already in discussion https://github.com/microsoft/vscode/issues/40233

**Comment by nazarhussain** [comment]
@nflaig @matthewkeil I used an extension called [Tabaqa](https://github.com/kalimahapps/tabaqa), which have it's own file `tabaqa.json`, on starting the VSCode, it copies over the settings from that file to the `settings.json` file, while preserving the existing content of `settings.json` file.

This way we can have team specific settings in the `tabaqa.json` and each team member can also have it's own configurations in `settings.json` file.

**Comment by nflaig** [comment]
>  on starting the VSCode, it copies over the settings from that file to the settings.json file

but doesn't this create a local diff if we track settings.json file? I don't see much of a downside putting a few extra settings in there that may be unused if you don't install the extension for it

**Comment by nazarhussain** [comment]
We are not tracking `settings.json` file, so anyone can use it on their likeness. We just keep `tabaqa.json` in VCS. 

This way you don't need to remember if you have local changes in `settings.json` you should not commit.

**Comment by nflaig** [comment]
> We just keep tabaqa.json in VCS

Ah got it, it's the other way around, not opposed to it other than we should consider if it's worth the extra extension

**Comment by nazarhussain** [comment]
> > We just keep tabaqa.json in VCS
> 
> Ah got it, it's the other way around, not opposed to it other than we should consider if it's worth the extra extension

We can and should use that extra extension until VSCode had this feature https://github.com/microsoft/vscode/issues/40233

File: .vscode/extensions.json [hunk 3]
```diff
@@ -1,6 +1,7 @@
 {
   "recommendations": [
     "biomejs.biome",
-    "esbenp.prettier-vscode"
+    "esbenp.prettier-vscode",
+    "KalimahApps.tabaqa"
```
Comment (nflaig) [inline .vscode/extensions.json]: this does not really seem to be used, we should really be careful installing extensions like this

image

Comment (nazarhussain) [reply .vscode/extensions.json]: Yes that's a pretty new extension, out of necessity. I checked it's code before using it.

Comment (matthewkeil) [reply .vscode/extensions.json]: Im not a fan of this extension approach... Everyone will need to have that and similar to what nico mentioned they have the same access as vscode which on my machine is a lot.

I think we should default to a few basic `setting.json` fields that are necessary.

I honestly even prefer that we do not commit a `settings.json` at all and instead commit `recommended.settings.json` and `recommended.extensions.json`.  Anyone that has repo issues and knows what they are doing will go into that folder and see them and will know what to do with them.

And that way we can all keep our own `settings.json` files intact and so can other users

**Comment by matthewkeil** [comment]
> ~~We can and should use that extra extension until~~ VSCode had this feature [microsoft/vscode#40233](https://github.com/microsoft/vscode/issues/40233)

This is the best solution... we should just wait till its part of vscode and only suggest what should be used but not commit/enforce it or extensions that are unwanted

File: .vscode/extensions.json (hunk 3 above)
Comment (nflaig) [reply .vscode/extensions.json]: I agree with the first part of what @matthewkeil mentioned but I don't like the `recommended.` files, having essential extensions noted + popup seems fine if we keep it to **only** mandatory extensions.

As per https://github.com/microsoft/vscode/issues/40233, it seems like `settings.json` even if the feature is implemented should be tracked via git. We can adopt that approach already and if we don't want any extra stuff in there I can find a way to gitignore my local changes to it, although imo it does not hurt to have 1-2 settingsi n there which would require an extension, those are ignored anyways if you don't have the extension, there is no harm.

Comment (matthewkeil) [reply .vscode/extensions.json]: I suppose I can live with that if its only biome and prettier.  Tabaqa and the settings file should not be included in this PR

File: .vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
+{
+  "window.title": "${activeEditorShort}${separator}${rootName}${separator}${profileName}${separator}[${activeRepositoryBranchName}]",
+  "editor.defaultFormatter": "esbenp.prettier-vscode",
+  // If you are using `sysoev.vscode-open-in-github` extension, make sure to enable this setting
+  // "openInGitHub.defaultBranch": "unstable",
```
Comment (nflaig) [inline .vscode/settings.json]: why is this commented out? If I enable it I still have this file modified locally which is not good

Comment (nflaig) [reply .vscode/settings.json]: these settings are just ignored if the extension is not installed, so there is no difference

File: .vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,22 @@
+{
+  "window.title": "${activeEditorShort}${separator}${rootName}${separator}${profileName}${separator}[${activeRepositoryBranchName}]",
+  "editor.defaultFormatter": "esbenp.prettier-vscode",
+  // To use this settings please install `sysoev.vscode-open-in-github` extension
```
Comment (nflaig) [inline .vscode/settings.json]: should we just put these at the end of the file?

```suggestion
  // For `sysoev.vscode-open-in-github` extension
```

File: .vscode/settings.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
+{
+  "window.title": "${activeEditorShort}${separator}${rootName}${separator}${profileName}${separator}[${activeRepositoryBranchName}]",
+  "editor.defaultFormatter": "esbenp.prettier-vscode",
+  // For `sysoev.vscode-open-in-github` extension
+  "openInGitHub.defaultBranch": "unstable",
```
Comment (nflaig) [inline .vscode/settings.json:5]: As noted in previous comment, not a fan of putting this in the middle of `editor.` settings but I guess it's fine

**Review by nflaig — APPROVED** [review]

File: .vscode/settings.json (hunk 6 above)
Comment (matthewkeil) [reply .vscode/settings.json:5]: LOL... its funny that this PR had more discussion than a lot

**Review by matthewkeil — APPROVED** [review]
Thanks for the thoughtful consideration during the PR review.  Much appreciated.  LGTM 🎸

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.23.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.23.0) :tada:

**Pull request merged**